pkg/il/interpreter/interpreterRun.got

```diff
@@ -0,0 +1,756 @@
+#define ERRF(...) tErr = fmt.Errorf(__VA_ARGS__); goto RETURN_ERR;
+
+#ifdef USE_GUARDS
+  #define STACK_OVERFLOW_GUARD(i) if sp > opStackSize - i { goto STACK_OVERFLOW };
+  #define STACK_UNDERFLOW_GUARD(i) if sp < i { goto STACK_UNDERFLOW };
+  #define HEAP_ACCESS_GUARD(i) if i >= hp { goto INVALID_HEAP_ACCESS };
+  #define HEAP_OVERFLOW_GUARD if hp == heapSize - 1 { goto HEAP_OVERFLOW };
+  #define STACK_OVERFLOW_BLOCK STACK_OVERFLOW: ERR("stack overflow");
+  #define STACK_UNDERFLOW_BLOCK STACK_UNDERFLOW: ERR("stack underflow");
+  #define INVALID_HEAP_ACCESS_BLOCK  INVALID_HEAP_ACCESS: ERR("invalid heap access")
+  #define HEAP_OVERFLOW_BLOCK HEAP_OVERFLOW: ERR("heap overflow");
+#else
+  #define STACK_OVERFLOW_GUARD(...)
+  #define STACK_UNDERFLOW_GUARD(...)
+  #define HEAP_ACCESS_GUARD(...)
+  #define HEAP_OVERFLOW_GUARD
+  #define STACK_OVERFLOW_BLOCK
+  #define STACK_UNDERFLOW_BLOCK
+  #define INVALID_HEAP_ACCESS_BLOCK
+  #define HEAP_OVERFLOW_BLOCK
+#endif
+
+#define LOAD_OP_CODE(target) \
+      target = body[ip]; \
+      ip++;
+
+#define LOAD_OP_CODE2(target1, target2) \
+      target1 = body[ip]; \
+      target2 = body[ip + 1]; \
+      ip = ip + 2;
+
+#define SET_REGISTER(index, value) \
+      registers[index] = value
+
+#define GET_REGISTER(index, target) \
+      target = registers[index]
+
+#define REGISTER(id) registers[id]
+
+#define STACK_REMOVE() \
+      sp--;
+
+#define STACK_REMOVE2() \
+      sp = sp - 2;
+
+#define STACK_POP(target) \
+      sp--; \
+      target = opstack[sp];
+
+#define STACK_PEEK(target) \
+      target = opstack[sp-1];
+
+#define STACK_PEEK2(t1, t2) \
+      t1 = opstack[sp-1]; \
+      t2 = opstack[sp-2];
+
+#define STACK_POP2(t1, t2) \
+      STACK_PEEK2(t1, t2) \
+      sp = sp - 2;
+
+#define STACK_PUSH(value) \
+      opstack[sp] = value; \
+      sp++;
+
+#define STACK_PUSH2(t1, t2) \
+      opstack[sp] = t1; \
+      opstack[sp + 1] = t2; \
+      sp = sp + 2;
+
+#define GET_HEAP_VALUE(index, target) \
+  HEAP_ACCESS_GUARD(index) \
+  target = heap[index];
+
+#define NEW_HEAP_VALUE(value, target) \
+  HEAP_OVERFLOW_GUARD \
+  target = hp; \
+  heap[hp] = value; \
+  hp++;
+
+#define U64(i0, i1) (uint64(i0) + uint64(i1)<<32)
+
+
+
+func (in *Interpreter) run(fn *il.Function, bag attribute.Bag, step bool) (Result, error) {
+
+	var registers [registerCount]uint32
+	var sp uint32 // stack-top pointer
+	var ip uint32 // instruction pointer
+	var fp uint32 // frame0top pointer
+
+	var opstack []uint32    // operand stack
+	var frames []stackFrame // call frames
+
+	var heap []interface{} // heap
+	var hp uint32          // heap-top pointer
+
+	// Initialize locals
+	strings := in.program.Strings()
+	body := in.code
+
+	// Temporaries
+	var code uint32
+	var t1 uint32
+	var t2 uint32
+	var t3 uint32
+	var ti64 int64
+	var tu64 uint64
+	var tf64 float64
+	var tVal interface{}
+	var tStr string
+	var tBool bool
+	var tFound bool
+	var tErr error
+
+	opstack = make([]uint32, opStackSize)
```

Comment:
That is what I started with. :) Then I tried using slices and it turned out to be faster. :)
I will run the benchmark again with arrays to double check and will update if it was only my memory playing tricks on me and arrays were indeed faster.